sdk/python/feast/infra/compute_engines/ray/job.py

```diff
@@ -0,0 +1,297 @@
+                    "Dataset should not be None after execution"
+                )
+                self._result_arrow = self._result_dataset.to_pandas().to_arrow()
+
+        # Handle validation if provided
+        if validation_reference:
+            try:
+                df = self._result_arrow.to_pandas()
+                validation_result = validation_reference.profile.validate(df)
+                if not validation_result.is_success:
+                    raise ValidationFailed(validation_result)
+            except ImportError:
+                logger.warning("DQM profiler not available, skipping validation")
+            except Exception as e:
+                logger.error(f"Validation failed: {e}")
+                raise ValueError(f"Data validation failed: {e}")
+
+        return self._result_arrow
+
+    def to_remote_storage(self) -> list[str]:
+        """Write the result to remote storage."""
+        if not self._config.batch_engine.staging_location:
+            raise ValueError("Staging location must be set for remote storage")
+
+        try:
+            self._ensure_executed()
+            assert self._result_dataset is not None, (
+                "Dataset should not be None after execution"
+            )
+            output_uri = (
+                f"{self._config.batch_engine.staging_location}/{str(uuid.uuid4())}"
+            )
+            self._result_dataset.write_parquet(output_uri)
+            logger.debug(f"Wrote result to {output_uri}")
+            return [output_uri]
+        except Exception as e:
+            raise RuntimeError(f"Failed to write to remote storage: {e}")
+
+    def persist(
+        self,
+        storage: SavedDatasetStorage,
+        allow_overwrite: bool = False,
+        timeout: Optional[int] = None,
+    ) -> str:
+        """Persist the result to the specified storage."""
+        if not isinstance(storage, SavedDatasetFileStorage):
+            raise ValueError(
+                f"Ray compute engine only supports SavedDatasetFileStorage, got {type(storage)}"
+            )
+
+        destination_path = storage.file_options.uri
+
+        # Check if destination already exists
+        if not destination_path.startswith(("s3://", "gs://", "hdfs://")):
+            import os
+
+            if not allow_overwrite and os.path.exists(destination_path):
+                raise SavedDatasetLocationAlreadyExists(location=destination_path)
+            os.makedirs(os.path.dirname(destination_path), exist_ok=True)
+
+        try:
+            self._ensure_executed()
+            assert self._result_dataset is not None, (
+                "Dataset should not be None after execution"
+            )
+            self._result_dataset.write_parquet(destination_path)
+            return destination_path
+        except Exception as e:
+            raise RuntimeError(f"Failed to persist dataset to {destination_path}: {e}")
+
+    def to_sql(self) -> str:
```

Comment:
Ray doesn't support this natively, so probably we should raise that?

But apparently there is this: https://github.com/datafusion-contrib/ray-sql

which is worth some investigation.